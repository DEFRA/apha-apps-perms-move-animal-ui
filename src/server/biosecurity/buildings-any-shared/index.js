import { QuestionPageController } from '../../common/controller/question-page-controller/question-page-controller.js'
import { BuildingsAnySharedAnswer } from '../../common/model/answer/buildings-any-shared/buildings-any-shared.js'
import { QuestionPage } from '../../common/model/page/question-page-model.js'
import { buildingsHowMinimiseContaminationPage } from '../buildings-how-minimise-contamination/index.js'
import { peopleDisinfectionPage } from '../people-disinfection/index.js'

export class BuildingsAnySharedPage extends QuestionPage {
  urlPath = '/biosecurity/buildings-any-shared'
  sectionKey = 'biosecurity'
  question =
    'Will the cattle share any buildings and equipment with the resident herd?'

  questionKey = 'buildingsAnyShared'
  Answer = BuildingsAnySharedAnswer

  /** @param {BuildingsAnySharedAnswer} answer */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  nextPage(answer) {
    if (answer.value === 'yes') {
      return buildingsHowMinimiseContaminationPage
    }
    return peopleDisinfectionPage
  }
}

export const buildingsAnySharedPage = new BuildingsAnySharedPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const buildingsAnyShared = new QuestionPageController(
  buildingsAnySharedPage
).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
