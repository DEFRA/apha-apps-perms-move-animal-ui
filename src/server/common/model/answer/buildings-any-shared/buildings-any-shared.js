import { RadioButtonAnswer } from '../radio-button/radio-button.js'

/** @import {RadioButtonConfig} from '../radio-button/radio-button.js' */

/**
 * export @typedef {'yes'|'no'} BuildingsAnySharedData
 * @typedef {{ animalsHoused: BuildingsAnySharedData }} BuildingsAnySharedPayload
 */

/**
 * @augments {RadioButtonAnswer<BuildingsAnySharedPayload>}
 */
export class BuildingsAnySharedAnswer extends RadioButtonAnswer {
  /** @type {RadioButtonConfig} */
  static config = {
    payloadKey: 'animalsHoused',
    hint: 'This includes shared spaces such as pens',
    options: {
      yes: { label: 'Yes' },
      no: { label: 'No' }
    },
    validation: {
      empty: 'Select if the animals will be housed'
    },
    layout: 'inline'
  }
}
