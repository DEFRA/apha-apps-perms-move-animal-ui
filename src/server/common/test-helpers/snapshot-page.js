import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
