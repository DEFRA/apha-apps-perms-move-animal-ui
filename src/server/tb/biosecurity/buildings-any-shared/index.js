import { TbQuestionPageController } from '../../question-page-controller.js'
import { BuildingsAnySharedAnswer } from '../../../common/model/answer/buildings-any-shared/buildings-any-shared.js'
import { QuestionPage } from '../../../common/model/page/question-page-model.js'
import { buildingsHowMinimiseContaminationPage } from '../buildings-how-minimise-contamination/index.js'
import { peopleDisinfectionPage } from '../people-disinfection/index.js'

export class BuildingsAnySharedPage extends QuestionPage {
  urlPath = '/biosecurity/buildings-any-shared'
  sectionKey = 'biosecurity'
  question = 'Will the incoming animals be housed?'

  questionKey = 'animalsHoused'
  Answer = BuildingsAnySharedAnswer

  /** @param {BuildingsAnySharedAnswer} answer */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  nextPage(answer) {
    if (answer.value === 'yes') {
      return buildingsHowMinimiseContaminationPage
    }
    return peopleDisinfectionPage
  }
}

export const buildingsAnySharedPage = new BuildingsAnySharedPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const buildingsAnyShared = new TbQuestionPageController(
  buildingsAnySharedPage
).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
