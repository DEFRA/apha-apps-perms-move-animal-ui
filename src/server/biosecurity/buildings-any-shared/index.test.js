import { BuildingsAnySharedAnswer } from '../../common/model/answer/buildings-any-shared/buildings-any-shared.js'
import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'
import { buildingsAnySharedPage, BuildingsAnySharedPage } from './index.js'
import { peopleDisinfectionPage } from '../people-disinfection/index.js'
import { buildingsHowMinimiseContaminationPage } from '../buildings-how-minimise-contamination/index.js'

const sectionKey = 'biosecurity'
const question =
  'Will the incoming cattle share any buildings and equipment with the resident herd?'
const questionKey = 'buildingsAnyShared'
const view = 'common/model/page/question-page.njk'
const pageUrl = '/biosecurity/buildings-any-shared'

describe('BuildingsAnySharedPage', () => {
  const page = new BuildingsAnySharedPage()

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(BuildingsAnySharedAnswer)
  })

  it('nextPage should return how to minimise contamination page when answer is "yes"', () => {
    const answer = new BuildingsAnySharedAnswer({ buildingsAnyShared: 'yes' })
    const nextPage = page.nextPage(answer)
    expect(nextPage).toBe(buildingsHowMinimiseContaminationPage)
  })

  it('nextPage should return people disinfection page when answer is "no"', () => {
    const answer = new BuildingsAnySharedAnswer({ buildingsAnyShared: 'no' })
    const nextPage = page.nextPage(answer)
    expect(nextPage).toBe(peopleDisinfectionPage)
  })

  it('should export page', () => {
    expect(buildingsAnySharedPage).toBeInstanceOf(BuildingsAnySharedPage)
  })

  describePageSnapshot({
    describes: 'buildingsAnySharedPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
