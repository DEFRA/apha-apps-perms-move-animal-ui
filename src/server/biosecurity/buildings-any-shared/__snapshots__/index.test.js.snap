// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildingsAnySharedPage buildingsAnySharedPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset" aria-describedby="buildingsAnyShared-hint">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      <h1 class="govuk-fieldset__heading">
        Will the incoming cattle share any buildings and equipment with the resident herd?
      </h1>
    </legend>
    <div id="buildingsAnyShared-hint" class="govuk-hint">
      This includes shared spaces such as pens
    </div>
    <div class="govuk-radios govuk-radios--inline" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="buildingsAnyShared" name="buildingsAnyShared" type="radio" value="yes">
        <label class="govuk-label govuk-radios__label" for="buildingsAnyShared">
          Yes
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="no" name="buildingsAnyShared" type="radio" value="no">
        <label class="govuk-label govuk-radios__label" for="no">
          No
        </label>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
